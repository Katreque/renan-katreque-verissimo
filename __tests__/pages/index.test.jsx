import React from 'react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'whatwg-fetch';

import HomePage from '../../pages/index';

const server = setupServer();
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('Home', () => {
    it('Deve mostrar o Spinner enquanto aguarda a resposta da API e removê-lo após o retorno.', async () => {
        server.use(
            rest.get(process.env.NEXT_PUBLIC_API, (req, res, ctx) => {
                return res(ctx.status(200), ctx.json({ msg: 'Mock' }));
            })
        );

        render(<HomePage />);

        const spinner = screen.getByTestId('home-spinner');
        expect(spinner).toBeInTheDocument();

        await waitFor(
            () => {
                expect(spinner).not.toBeInTheDocument();
                expect(screen.getByText('Welcome to Next.js!')).toBeVisible();
            },
            { timeout: 3000 }
        );
    });

    it('Deve manter o Spinner e mostrar os textos correspondentes caso a API retorne algum erro.', async () => {
        server.use(
            rest.get(process.env.NEXT_PUBLIC_API, (req, res, ctx) => {
                return res(ctx.status(500), ctx.json({ msg: 'Mock' }));
            })
        );

        render(<HomePage />);

        const spinner = screen.getByTestId('home-spinner');
        expect(spinner).toBeInTheDocument();

        await waitFor(
            () => {
                expect(spinner).toBeVisible();
                expect(
                    screen.getByText('Houston, we have a problem...')
                ).toBeVisible();
                expect(
                    screen.getByText('Recarregue a página e tente novamente.')
                ).toBeVisible();
            },
            { timeout: 3000 }
        );
    });
});
