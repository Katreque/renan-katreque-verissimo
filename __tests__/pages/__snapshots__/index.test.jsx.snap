// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Snapshot do componente após o carregamento da API e antes de iniciar o jogo. 1`] = `
<DocumentFragment>
  .c0 {
  height: 100px;
  margin: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  color: white;
  background-color: darkcyan;
  font-size: 1.5em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid green;
  border-radius: 3px;
  cursor: pointer;
}

<div
    class="c0"
  >
    <button
      class="c1"
    >
      Iniciar o Jogo
    </button>
  </div>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  row-gap: 20px;
  -webkit-column-gap: 5px;
  column-gap: 5px;
  margin-bottom: 20px;
}

<div
    class="c0"
  />
</DocumentFragment>
`;

exports[`HomePage Snapshot do componente caso a API retorne algum erro 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-animation-name: fCDSQb;
  animation-name: fCDSQb;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
  margin-top: 50px;
  font-size: 1.25em;
  -webkit-animation-name: fCDSQb;
  animation-name: fCDSQb;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-delay: 500ms;
  animation-delay: 500ms;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c3 {
  margin-bottom: 10px;
  font-weight: bold;
}

.c0 {
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <div
      class="c1"
      data-testid="spinner-3"
    >
      <span
        style="background-color: rgb(245, 33, 50); width: 20px; height: 20px; border-radius: 100%; display: inline-block; position: relative; animation-fill-mode: both; animation: react-spinners-RotateLoader-rotate 5s 0s infinite cubic-bezier(0.7, -0.13, 0.22, 0.86);"
      >
        <span
          style="background-color: rgb(245, 33, 50); width: 20px; height: 20px; border-radius: 100%; opacity: 0.8; position: absolute; top: 0px; left: -28px;"
        />
        <span
          style="background-color: rgb(245, 33, 50); width: 20px; height: 20px; border-radius: 100%; opacity: 0.8; position: absolute; top: 0px; left: 28px;"
        />
      </span>
    </div>
    <div
      class="c2"
    >
      <span
        class="c3"
      >
        Houston, we have a problem...
      </span>
      <small>
        Recarregue a página e tente novamente.
      </small>
    </div>
  </div>
</DocumentFragment>
`;

exports[`HomePage Snapshot do componente depois de iniciar o jogo e selecionar uma carta. 1`] = `
<DocumentFragment>
  .c0 {
  height: 100px;
  margin: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  font-size: 1.5em;
  font-weight: 700;
}

<div
    class="c0"
  >
    <p
      class="c1"
    >
      Selecione uma carta:
    </p>
  </div>
  .c7 {
  -webkit-animation: kcCFSJ;
  animation: kcCFSJ;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c7.close-modal {
  opacity: 0;
  visibility: collapse;
}

.c13 {
  z-index: 3;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(84,155,247,0.5);
  -webkit-animation-delay: 0.5s;
  animation-delay: 0.5s;
}

.c8 {
  z-index: 4;
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  width: 600px;
  max-width: 100%;
  max-height: 100%;
  background-color: white;
  border-radius: 20px;
  -webkit-animation-delay: 0.75s;
  animation-delay: 0.75s;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 10px;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  margin: 10px;
  padding: 5px;
}

.c12 {
  margin: 10px;
}

.c1 {
  width: 150px;
  height: 300px;
}

.c1:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
  -webkit-transition: scale 200ms;
  transition: scale 200ms;
}

.c2 {
  position: relative;
}

.c3 {
  background-image: url(https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png);
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  -webkit-animation: ddxGlr;
  animation: ddxGlr;
  -webkit-animation-delay: 0ms;
  animation-delay: 0ms;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c14 {
  background-image: url(https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png);
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  -webkit-animation: fCDSQb;
  animation: fCDSQb;
  -webkit-animation-delay: 20ms;
  animation-delay: 20ms;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c16 {
  background-image: url(https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png);
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  -webkit-animation: fCDSQb;
  animation: fCDSQb;
  -webkit-animation-delay: 40ms;
  animation-delay: 40ms;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c5 {
  background-image: url(https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/162x341/arcano3.jpg);
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  -webkit-animation: ddxGlr;
  animation: ddxGlr;
  -webkit-animation-delay: 0ms;
  animation-delay: 0ms;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c4 {
  position: absolute;
  width: 100%;
  height: 100%;
  -webkit-animation-duration: 0.75s;
  animation-duration: 0.75s;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  -webkit-perspective: 1000px;
  -moz-perspective: 1000px;
  -ms-perspective: 1000px;
  perspective: 1000px;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
}

.c6 {
  position: absolute;
  width: 100%;
  height: 100%;
  -webkit-animation-duration: 0.75s;
  animation-duration: 0.75s;
  -webkit-animation-direction: reverse;
  animation-direction: reverse;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  -webkit-perspective: 1000px;
  -moz-perspective: 1000px;
  -ms-perspective: 1000px;
  perspective: 1000px;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
}

.c15 {
  cursor: pointer;
  z-index: 2;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  row-gap: 20px;
  -webkit-column-gap: 5px;
  column-gap: 5px;
  margin-bottom: 20px;
}

<div
    class="c0"
  >
    <div
      class="c1 c2"
      data-testid="5-0"
    >
      <div
        class="c1 c3 c4"
        mode="both"
      />
      <div
        class="c1 c5 c6"
        direction="reverse"
        mode="both"
      />
    </div>
    <div
      data-testid="card-expanded"
    >
      <div
        class="c7 c8"
      >
        <div
          class="c9"
        >
          <div
            class="c10"
          >
            <img
              alt="A IMPERATRIZ"
              data-nimg="1"
              decoding="async"
              height="300"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fdkw5ssdvaqf8l.cloudfront.net%2Fstatic%2Fpsr%2Fbr%2Fframework%2Fyii%2Fimages%2Fcontent%2Fpt-br%2Fproduct%2Ftarot%2Fmarselha%2F162x341%2Farcano3.jpg&w=384&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fdkw5ssdvaqf8l.cloudfront.net%2Fstatic%2Fpsr%2Fbr%2Fframework%2Fyii%2Fimages%2Fcontent%2Fpt-br%2Fproduct%2Ftarot%2Fmarselha%2F162x341%2Farcano3.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fdkw5ssdvaqf8l.cloudfront.net%2Fstatic%2Fpsr%2Fbr%2Fframework%2Fyii%2Fimages%2Fcontent%2Fpt-br%2Fproduct%2Ftarot%2Fmarselha%2F162x341%2Farcano3.jpg&w=384&q=75 2x"
              style="color: transparent;"
              width="150"
            />
          </div>
          <div
            class="c11"
          >
            <h2>
              A IMPERATRIZ
            </h2>
            <p>
              Mussum Ipsum, cacilds vidis litro abertis. Viva Forevis aptent taciti sociosqu ad litora torquent. Admodum accumsan disputationi eu sit. Vide electram sadipscing et per.
            </p>
            <button
              class="c12"
            >
              Mais uma?
            </button>
          </div>
        </div>
      </div>
      <div
        class="c7 c13"
      />
    </div>
    <div
      class="c1 c14 c15"
      data-testid="4-1"
      mode="both"
    />
    <div
      class="c1 c16 c15"
      data-testid="4-2"
      mode="both"
    />
  </div>
</DocumentFragment>
`;

exports[`HomePage Snapshot do componente depois de iniciar o jogo. 1`] = `
<DocumentFragment>
  .c0 {
  height: 100px;
  margin: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  font-size: 1.5em;
  font-weight: 700;
}

<div
    class="c0"
  >
    <p
      class="c1"
    >
      Selecione uma carta:
    </p>
  </div>
  .c1 {
  width: 150px;
  height: 300px;
}

.c1:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
  -webkit-transition: scale 200ms;
  transition: scale 200ms;
}

.c2 {
  background-image: url(https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png);
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  -webkit-animation: fCDSQb;
  animation: fCDSQb;
  -webkit-animation-delay: 0ms;
  animation-delay: 0ms;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c4 {
  background-image: url(https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png);
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  -webkit-animation: fCDSQb;
  animation: fCDSQb;
  -webkit-animation-delay: 20ms;
  animation-delay: 20ms;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c5 {
  background-image: url(https://dkw5ssdvaqf8l.cloudfront.net/static/psr/br/framework/yii/images/content/pt-br/product/tarot/marselha/back-blue-card.png);
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  -webkit-animation: fCDSQb;
  animation: fCDSQb;
  -webkit-animation-delay: 40ms;
  animation-delay: 40ms;
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}

.c3 {
  cursor: pointer;
  z-index: 2;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  row-gap: 20px;
  -webkit-column-gap: 5px;
  column-gap: 5px;
  margin-bottom: 20px;
}

<div
    class="c0"
  >
    <div
      class="c1 c2 c3"
      data-testid="4-0"
      mode="both"
    />
    <div
      class="c1 c4 c3"
      data-testid="4-1"
      mode="both"
    />
    <div
      class="c1 c5 c3"
      data-testid="4-2"
      mode="both"
    />
  </div>
</DocumentFragment>
`;

exports[`HomePage Snapshot do componente enquanto aguarda a resposta da API. 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-animation-name: fCDSQb;
  animation-name: fCDSQb;
  -webkit-animation-duration: 500ms;
  animation-duration: 500ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c0 {
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <div
      class="c1"
      data-testid="spinner-1"
    >
      <span
        style="background-color: rgb(75, 131, 166); width: 20px; height: 20px; border-radius: 100%; display: inline-block; position: relative; animation-fill-mode: both; animation: react-spinners-RotateLoader-rotate 2s 0s infinite cubic-bezier(0.7, -0.13, 0.22, 0.86);"
      >
        <span
          style="background-color: rgb(75, 131, 166); width: 20px; height: 20px; border-radius: 100%; opacity: 0.8; position: absolute; top: 0px; left: -28px;"
        />
        <span
          style="background-color: rgb(75, 131, 166); width: 20px; height: 20px; border-radius: 100%; opacity: 0.8; position: absolute; top: 0px; left: 28px;"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
